packages/medusa-plugin-add-ons/src/services/add-on.js

```diff
@@ -0,0 +1,170 @@
+import _ from "lodash";
+import { BaseService } from "medusa-interfaces";
+import { Validator, MedusaError } from "medusa-core-utils";
+
+class AddOnService extends BaseService {
+  static Events = {
+    UPDATED: "add_on.updated",
+    CREATED: "add_on.created",
+  };
+
+  constructor(
+    {
+      addOnModel,
+      productService,
+      productVariantService,
+      regionService,
+      eventBusService,
+    },
+    options
+  ) {
+    super();
+
+    this.addOnModel_ = addOnModel;
+
+    this.productService_ = productService;
+
+    this.productVariantService_ = productVariantService;
+
+    this.regionService_ = regionService;
+
+    this.eventBus_ = eventBusService;
+
+    this.options_ = options;
+  }
+
+  /**
+   * Used to validate add-on ids. Throws an error if the cast fails
+   * @param {string} rawId - the raw add-on id to validate.
+   * @return {string} the validated id
+   */
+  validateId_(rawId) {
+    const schema = Validator.objectId();
+    const { value, error } = schema.validate(rawId.toString());
+    if (error) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "The addOnId could not be casted to an ObjectId"
+      );
+    }
+
+    return value;
+  }
+
+  /**
+   * @param {Object} selector - the query object for find
+   * @return {Promise} the result of the find operation
+   */
+  list(selector, offset, limit) {
+    return this.addOnModel_.find(selector, {}, offset, limit);
+  }
+
+  /**
+   * Gets an add-on by id.
+   * @param {string} addOnId - the id of the add-on to get.
+   * @return {Promise<AddOn>} the add-on document.
+   */
+  async retrieve(addOnId) {
+    const validatedId = this.validateId_(addOnId);
+    const addOn = await this.addOnModel_
+      .findOne({ _id: validatedId })
+      .catch((err) => {
+        throw new MedusaError(MedusaError.Types.DB_ERROR, err.message);
+      });
+
+    if (!addOn) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Add-on with ${addOnId} was not found`
+      );
+    }
+    return addOn;
+  }
+
+  /**
+   * Creates an add-on.
+   * @param {object} addOn - the add-on to create
+   * @return {Promise} resolves to the creation result.
+   */
+  async create(addOn) {
+    return this.addOnModel_
+      .create(addOn)
+      .then((result) => {
+        this.eventBus_.emit(AddOnService.Events.CREATED, result);
+        return result;
+      })
+      .catch((err) => {
+        throw new MedusaError(MedusaError.Types.DB_ERROR, err.message);
+      });
+  }
+
+  /**
+   * Updates an add-on. Metadata updates should use dedicated methods, e.g.
+   * `setMetadata`, etc. The function will throw errors if metadata updates
+   * are attempted.
+   * @param {string} addOnId - the id of the add-on. Must be a string that
+   *   can be casted to an ObjectId
```

Comment:
Maybe check if products exist in `valid_for`

Reply:
In what context? You need to be able to update the valid products for a given add-on, so we can't really add a check for that

Reply:
My thought was just that in order to not pollute the `valid_for` array with a bunch of ids for products that might not exist we should do the following:

If the `valid_for` array is present on `addOn` we go through each and make sure that the product id in the array is associated with an existing product if it is we go ahead and create the add on if not we throw an invalid data error.

Reply:
Ahh alright I see. I'll add it!